[DataGrid] Add error handling (#134)
mui-org/material-ui-x: packages/storybook/src/documentation/pages/error-handling.stories.mdx

import { XGrid } from '@material-ui/x-grid';
import ExceptionDemo from './demos/error/exception.demo';
import ServerErrorDemo from './demos/error/serverError.demo';

<Meta title="1. Docs/13. Error handling" component={XGrid} />

# Error handling

Handling error on the UI and in the grid, is critical to avoid breaking the whole application, or even displaying corrupted data.

> Therefore, XGrid is wrapped with an error boundary internally. So if the component throw an unhandled exception anywhere in the tree,

Reviewer: ```suggestion
Therefore, XGrid is wrapped with an [error boundary](https://reactjs.org/docs/error-boundaries.html) internally. So if the component throws an unhandled exception anywhere in the tree,
```

the exception will get caught by the error boundary, and a nice error message overlay will be displayed.

<ExceptionDemo />

## Show server errors

Server errors can be displayed using the `showError` method of the `ApiRef` as below.

```tsx
const apiRef = useApiRef();

> useEffect(()=> {

Reviewer: ```suggestion
React.useEffect(()=> {
```

    service.loadBadRows().catch((error)=> {

>         apiRef.current.showError({message: 'Error loading rows.'});

Reviewer: I'm not sure about the imperative API, it feels like a demo with the declarative API would be better. Especially to recover from errors.

    }
}, [apiRef]);

<XGrid rows={rows} columns={columns} apiRef={apiRef} />

```

<ServerErrorDemo />

> ## Customising the error component
> 
> ## Events and OnError option

Reviewer: ```suggestion
## Customising the error component

## Events and onError option

### Error reporting tools

https://www.smashingmagazine.com/2020/06/react-error-handling-reporting-error-boundary-sentry/
```

import { XGrid } from '@material-ui/x-grid';
import ExceptionDemo from './demos/error/exception.demo';
import ServerErrorDemo from './demos/error/serverError.demo';
import ServerErrorStateDemo from './demos/error/serverErrorState.demo';
import CustomErrorDemo from './demos/error/customError.demo';

<Meta title="1. Docs/13. Error handling" component={XGrid} />

# Error handling

Handling error on the UI and in the grid, is critical to avoid breaking the whole application, or even displaying corrupted data.
Therefore, XGrid is wrapped with an [error boundary](https://reactjs.org/docs/error-boundaries.html) internally. So if the component throws an unhandled exception anywhere in the tree,
the exception will get handled by the error boundary, and a nice error message overlay will be displayed.

<ExceptionDemo />

## Customising the error component

The error component is also part of the customization components and can be replaced as below.

```tsx
...
return (
    <XGrid
      rows={rows}
      columns={columns}
      components={{
        errorOverlay: CustomErrorOverlay,
      }}
    />
  );
```

<CustomErrorDemo />

## Show server errors

### Using state

```tsx

> const [error, setError] = React.useState();
> React.useEffect(()=> {
>     service.loadBadRows().catch((error)=> {
>         setError({ message: `An error occured! Details: ${error.message}` });
>     }
> }, [setError]);
> 
> <XGrid rows={rows} columns={columns} error = {error} />
> 

Reviewer: ```suggestion
const [error, setError] = React.useState();

React.useEffect(()=> {
  service.loadBadRows().catch((error)=> {
    setError({ message: `An error occured! Details: ${error.message}` });
  }
}, [setError]);

<XGrid rows={rows} columns={columns} error={error} />
```

```

<ServerErrorStateDemo />

### Using the ApiRef

Server errors can be displayed using the `showError` method of the `ApiRef` as below.

```tsx

> const apiRef = useApiRef();
> 
> React.useEffect(()=> {
>     service.loadBadRows().catch((error)=> {
>         apiRef.current.showError({message: 'Error loading rows.'});
>     }
> }, [apiRef]);
> 
> <XGrid rows={rows} columns={columns} apiRef={apiRef} />
> 

Reviewer: ```suggestion
const apiRef = useApiRef();

React.useEffect(()=> {
  service.loadBadRows().catch((error)=> {
    apiRef.current.showError({ message: 'Error loading rows.' });
  }
}, [apiRef]);

<XGrid rows={rows} columns={columns} apiRef={apiRef} />
```
